crates/core_arch/src/aarch64/tbi.rs

```diff
@@ -0,0 +1,229 @@
+extern crate alloc;
+use crate::arch::asm;
+use alloc::boxed::Box;
+use core::convert::{AsMut, AsRef};
+use core::ops::{Deref, DerefMut, Drop};
+
+/// A TBI-enabled [Box].
+///
+/// This allows setting the top byte to arbitrary values via top-byte ignore (TBI).
+#[derive(Debug)]
+#[unstable(feature = "stdarch_aarch64_tbi", issue = "none")]
+pub struct TBIBox<T>(
+    Option<Box<T>>,
+    /* original pointer for deallocation: */ *mut T,
+    // Without keeping a copy of the top byte around, it is difficult to cleanly extract the top byte from the
+    // underlying pointer on-demand since `from_raw` requires consuming `self.0`.
+    /* top byte for easy retrieval: */
+    u8,
+);
+
+impl<T> TBIBox<T> {
+    fn construct_new_tbibox(b: Box<T>, top_byte: u8) -> TBIBox<T> {
+        // Pointer modificaton: this `Box` hasn't been exposed to the user yet, so the user cannot have a copy of the pointer with
+        // an unset top byte; we use this chance to set the top byte before they can see it.
+        let original_ptr: *mut T = Box::into_raw(b);
+        let ptr =
+            original_ptr.map_addr(|addr| (addr & 0x00ffffffffffffff) | ((top_byte as usize) << 56));
+        unsafe {
+            asm!(
+                "/* pretend that we're memcpy'ing from {original_ptr} to {ptr}... */",
+                original_ptr = in(reg) original_ptr,
+                ptr = in(reg) ptr
+            )
+        };
+        // Reconstruct the `Box` using the address with the new top byte and return that, wrapped as a TBIBox
+        Self(Some(unsafe { Box::from_raw(ptr) }), original_ptr, top_byte)
```

Comment:
This doesn't hide things sufficiently well from the compiler. The compiler can still see that `ptr` is just the same as `original_ptr` with an offset applied -- making this still UB.

You need the computation of `ptr` itself to occur in a "black box".

Reply:
Ah yes I see, maybe I could move the step that actually sets the top byte i.e. `addr | top_byte_shifted` into the inline asm block and do it there with an explicit `ORR`, do you think that would work?

Reply:
Yeah that would be better.